nsqd/protocol_v2.go

```diff
@@ -622,6 +622,10 @@ func (p *protocolV2) SUB(client *clientV2, params [][]byte) ([]byte, error) {
 			time.Sleep(1 * time.Millisecond)
 			continue
 		}
+		if channel.IsClientConnectionCountExceeded() {
```

Comment:
I think this check could be inside `channel.AddClient()` - that way the channel lock is only taken once, and you avoid some races:
  * multiple clients connect concurrently, both are added, then both check if limit exceeded, both get True even though only one should
  * not sure but maybe: client added, can get some messages? then disconnects because limit exceeded

so `Channel.AddClient()` could be adjusted to do this check and return an error.

Reply:
Good catch. Will do.

Reply:
> not sure but maybe: client added, can get some messages? then disconnects because limit exceeded

This should not happen. As this logic is all under `SUB` command callback. As an error raised in `SUB` command callback function, all io operations will not be further invoked.